api/envoy/config/listener/v3/listener.proto

```diff
@@ -141,21 +124,25 @@ message Listener {
   // A filter chain matcher resolving the filter chain name based on the connection properties.
   // The following inputs are supported by the matcher:
   //
-  // 1. :ref:`DestinationPort <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.DestinationPortInput>`.
-  // 2. :ref:`DestinationIP <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.DestinationIPInput>`.
-  // 3. :ref:`SNI <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SNIInput>`.
-  // 4. :ref:`TransportProtocol <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.TransportProtocolInput>`.
-  // 5. :ref:`ApplicationProtocol <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.ApplicationProtocolInput>`.
-  // 6. :ref:`DirectSourceIP <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.DirectSourceIPInput>`.
-  // 7. :ref:`SourceType <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SourceTypeInput>`.
-  // 8. :ref:`SourceIP <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SourceIPInput>`.
-  // 9. :ref:`SourcePort <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SourcePortInput>`.
+  // 1. :ref:`DestinationPort <envoy_v3_api_msg_config.listener.v3.DestinationPortInput>`.
+  // 2. :ref:`DestinationIP <envoy_v3_api_msg_config.listener.v3.DestinationIPInput>`.
+  // 3. :ref:`SNI <envoy_v3_api_msg_config.listener.v3.SNIInput>`.
+  // 4. :ref:`TransportProtocol <envoy_v3_api_msg_config.listener.v3.TransportProtocolInput>`.
+  // 5. :ref:`ApplicationProtocol <envoy_v3_api_msg_config.listener.v3.ApplicationProtocolInput>`.
+  // 6. :ref:`DirectSourceIP <envoy_v3_api_msg_config.listener.v3.DirectSourceIPInput>`.
+  // 7. :ref:`SourceType <envoy_v3_api_msg_config.listener.v3.SourceTypeInput>`.
+  // 8. :ref:`SourceIP <envoy_v3_api_msg_config.listener.v3.SourceIPInput>`.
+  // 9. :ref:`SourcePort <envoy_v3_api_msg_config.listener.v3.SourcePortInput>`.
+  //
+  // If specified, all :ref:`filter_chains <envoy_v3_api_field_config.listener.v3.Listener.filter_chains>` must
+  // have a non-empty :ref:`name <envoy_v3_api_field_config.listener.v3.FilterChain.name>` and empty :ref:`filter_chain_match <envoy_v3_api_msg_config.listener.v3.FilterChain.filter_chain_match>`.
   // [#not-implemented-hide:]
-  repeated common.matcher.v3.Matcher filter_chain_matcher = 31;
+  repeated xds.type.matcher.v3.Matcher filter_chain_matcher = 31;
 
-  // Filter chains.
+  // Defines the discovery service configuration for filter chains indexed by the filter chain name.
+  // The set of filter chain names must be disjoint from the filter chain names in the field :ref:`filter_chains <envoy_v3_api_field_config.listener.v3.Listener.filter_chains>`.
   // [#not-implemented-hide:]
-  repeated NamedFilterChain named_filter_chains = 32;
+  map<string, core.v3.ExtensionConfigSource> filter_chain_discovery = 32;
```

Comment:
I think we should still make it possible to inline a `FilterChain` into this map if a control plane doesn't want to use ECDS but also doesn't want to duplicate the `FilterChain` in multiple leaves of the tree.

Reply:
I think we should reuse the existing `filter_chains` field for concrete chains. The leaves are always just names. Duplicates can be rejected once the matcher is used. This is aligned with staging discovery for later.

Reply:
I think that approach could lead to a lot of wire-ineffeciency.  Even if we de-dup later, there's a lot of overhead involved in sending the same proto many times on the wire.

It's not clear to me what the down-side is of allowing inline filter chains to be accessed in this map.